### PR TITLE
BUG: signal.sepfir2d: reject filters too long for the input axis

### LINKED ISSUE #24681: BUG: signal._spline.sepfir2d: heap-buffer-overflow
### Describe your issue.

I saw the following ASAN error when running SciPy 1.17.1 under asan in our internal test environment.

```
signal/tests/test_bsplines.py::TestSepfir2d::test_sepfir2d_strided_2[numpy]
SanitizerError
AddressSanitizer: heap-buffer-overflow third_party/py/scipy/signal/_splinemodule.h:322:34 in void _fir_mirror_symmetric<double>(double*, double*, int, double*, int, int, int)
Details
=================================================================
==3014==ERROR: AddressSanitizer: heap-buffer-overflow on address 0x7c113d6339f8 at pc 0x7b5127772657 bp 0x7ffcee9ebd30 sp 0x7ffcee9ebd28
READ of size 8 at 0x7c113d6339f8 thread T0
    #0 0x7b5127772656 in void _fir_mirror_symmetric<double>(double*, double*, int, double*, int, int, int) third_party/py/scipy/signal/_splinemodule.h:322:34
    #1 0x7b51277718c9 in int _separable_2Dconvolve_mirror<double>(double*, double*, int, int, double*, double*, int, int, long*, long*) third_party/py/scipy/signal/_splinemodule.h:349:13
    #2 0x7b5127770264 in FIRsepsym2d(_object*, _object*) third_party/py/scipy/signal/_splinemodule.cc:100:11
    #3 0x55ad4ad574fb in cfunction_call third_party/python_runtime/v3_12/Objects/methodobject.c:548:18
    #4 0x55ad4acf0179 in _PyObject_MakeTpCall third_party/python_runtime/v3_12/Objects/call.c:240:18
    #5 0x55ad4acefc69 in _PyObject_VectorcallTstate third_party/python_runtime/v3_12/Include/internal/pycore_call.h:90:16
    #6 0x55ad4acf0faa in PyObject_Vectorcall third_party/python_runtime/v3_12/Objects/call.c:325:12
    #7 0x55ad4ae8ced6 in _PyEval_EvalFrameDefault Python/bytecodes.c:2715:19
    #8 0x55ad4ae7c6ae in _PyEval_EvalFrame third_party/python_runtime/v3_12/Include/internal/pycore_ceval.h:89:16
    #9 0x55ad4ae7c6ae in _PyEval_Vector third_party/python_runtime/v3_12/Python/ceval.c:1685:12
...
0x7c113d6339f8 is located 8 bytes before 128-byte region [0x7c113d633a00,0x7c113d633a80)
allocated by thread T0 here:
    #0 0x55ad4aab50a4 in malloc third_party/llvm/llvm-project/compiler-rt/lib/asan/asan_malloc_linux.cpp:87:3
    #1 0x7f51bdd07619 in _npy_alloc_cache third_party/py/numpy/_core/src/multiarray/alloc.c:139:9
    #2 0x7f51bdd07619 in default_malloc third_party/py/numpy/_core/src/multiarray/alloc.c:325:12
    #3 0x7f51bdd07896 in PyDataMem_UserNEW third_party/py/numpy/_core/src/multiarray/alloc.c:401:14
    #4 0x7f51bdd37106 in PyArray_NewFromDescr_int third_party/py/numpy/_core/src/multiarray/ctors.c:903:20
    #5 0x7f51bdd3d72c in PyArray_NewFromDescrAndBase third_party/py/numpy/_core/src/multiarray/ctors.c:1047:12
    #6 0x7f51bdd3d72c in PyArray_NewFromDescr third_party/py/numpy/_core/src/multiarray/ctors.c:1032:12
    #7 0x7f51bdd3d72c in PyArray_Empty_int third_party/py/numpy/_core/src/multiarray/ctors.c:3110:28
    #8 0x7f51bddb1c4a in array_empty third_party/py/numpy/_core/src/multiarray/multiarraymodule.c:2068:28
    #9 0x55ad4ad56711 in cfunction_vectorcall_FASTCALL_KEYWORDS third_party/python_runtime/v3_12/Objects/methodobject.c:438:24
    #10 0x7b513406e33c in __Pyx_PyObject_FastCallDict blaze-out/k8-fastbuild-asan/genfiles/third_party/py/numpy/_cython/random/_common.c:24343:20
    #11 0x7b5134085835 in __pyx_f_5numpy_6random_7_common_double_fill blaze-out/k8-fastbuild-asan/genfiles/third_party/py/numpy/_cython/random/_common.c:9350:19
    #12 0x7b5131a394a8 in __pyx_pf_5numpy_6random_6mtrand_11RandomState_18random_sample blaze-out/k8-fastbuild-asan/genfiles/third_party/py/numpy/_cython/random/mtrand.c:9816:15
    #13 0x7b5131a394a8 in __pyx_pw_5numpy_6random_6mtrand_11RandomState_19random_sample blaze-out/k8-fastbuild-asan/genfiles/third_party/py/numpy/_cython/random/mtrand.c:9786:13
    #14 0x7b5131a76d21 in __Pyx_CyFunction_Vectorcall_FASTCALL_KEYWORDS blaze-out/k8-fastbuild-asan/genfiles/third_party/py/numpy/_cython/random/mtrand.c:36107:12
    #15 0x55ad4acf2a8a in _PyObject_VectorcallTstate third_party/python_runtime/v3_12/Include/internal/pycore_call.h:92:11
    #16 0x55ad4acf2a8a in PyObject_VectorcallMethod third_party/python_runtime/v3_12/Objects/call.c:887:24
    #17 0x7b5131a42ad0 in __pyx_pf_5numpy_6random_6mtrand_11RandomState_38rand blaze-out/k8-fastbuild-asan/genfiles/third_party/py/numpy/_cython/random/mtrand.c:15162:19
...

SUMMARY: AddressSanitizer: heap-buffer-overflow third_party/py/scipy/signal/_splinemodule.h:322:34 in void _fir_mirror_symmetric<double>(double*, double*, int, double*, int, int, int) (//third_party/py/scipy:signal/tests/test_bsplines)
```

I asked Gemini to diagnose and fix the problem and it suggests:

"
The function originally implemented an optimized 3-loop strategy (first part, middle section, end boundary conditions) for avoiding inner loop branching, but lacked checks for overlap. Whenever Nhdiv2 * 2 >= N, the first part and end part boundary loops would overlap their output assignments (n - N + 1 <= Nhdiv2), resulting in two bugs:

out[n] is written out-of-bounds or overwritten.
For small n, the logic in end boundary conditions loops past the boundary and reads negative indices into inptr, resulting directly in the ASAN heap-buffer-overflow.
"

```
diff --git a/signal/_splinemodule.h b/signal/_splinemodule.h
--- a/signal/_splinemodule.h
+++ b/signal/_splinemodule.h
@@ -276,6 +276,27 @@ void _fir_mirror_symmetric(T *in, T *out
     T *inptr;
     T *hptr;

+    if (Nhdiv2 * 2 >= N) {
+        for (n = 0; n < N; n++) {
+            T sum = 0.0;
+            hptr = h;
+            for (k = -Nhdiv2; k <= Nhdiv2; k++) {
+                int idx = n - k;
+                while (idx < 0 || idx >= N) {
+                    if (idx < 0) {
+                        idx = -idx - 1;
+                    }
+                    if (idx >= N) {
+                        idx = 2 * N - 1 - idx;
+                    }
+                }
+                sum += (*hptr++) * in[idx * instride];
+            }
+            out[n * outstride] = sum;
+        }
+        return;
+    }
+
     /* first part boundary conditions */
     outptr = out;
     for (n=0; n < Nhdiv2; n++) {
```

and indeed, the buffer overflow goes away with this patch, but I don't understand this code at all so I'm not going to send that as a PR, per the AI policy.

### Reproducing Code Example

```python
signal/tests/test_bsplines.py::TestSepfir2d::test_sepfir2d_strided_2[numpy]
```

### Error message

```shell
See above.
```

### SciPy/NumPy/Python version and system information

```shell
This is a self-build of SciPy 1.17.1, NumPy 2.4.1, Python 3.12.
```

### PR DESCRIPTION
**Pull request by tekinertekin**
#### Reference issue

Closes gh-24681.

#### What does this implement/fix?
`_fir_mirror_symmetric` handles an out-of-bounds index only once, so it requires
`2*(len(filt)//2)` samples along the axis being filtered. On a shorter axis it read
past both ends of the input, and once `len(filt)//2` grew larger than the axis size
it also wrote outside the output; the values returned were whatever was stored in
the adjacent heap space. Validate both axes in `FIRsepsym2d` instead of relying on
the caller.

The bound is exact, not the `len(filt) <= n` the TODO suggests, where `n` is the
length of the filtered axis: `len(filt) == n + 1` is still in-bounds and matches a
mirror-symmetric reference convolution, so it stays accepted.

This also removes the ASAN suppression and the two xfail markers that were hiding
the problem. `test_sepfir2d_strided_2` only ever called `sepfir2d` with a filter
too long for its image, so it is replaced by `test_sepfir2d_filter_too_long`;
`test_sepfir2d_strided_3` keeps its valid half, and its expected dtype is now
pinned, which the xfail had been masking as well.

#### Additional information

The bound was checked four independent ways, all agreeing on
`N >= 2*(len(filt)//2)`:

| check | result |
|---|---|
| index-range sweep over `(N, len(filt))` | 54/54 agree; also shows the output overflow |
| ASAN on a verbatim copy of the function | 50/50 agree; clean at the boundary |
| all-ones 4x4 image, `len(filt) == 7` | correct answer 225; got `min=145`, and `min=17` on a rerun |
| vs. a mirror-symmetric reference convolution | identical for every in-bounds `(N, len(filt))` |

Because the results for the rejected inputs were undefined, there is no output any
caller could have relied on; the same call returned different values on reruns.

`scipy/signal/tests/test_bsplines.py` goes from 24 passed / 5 xfailed / 2 xpassed
to 39 passed with no xfails. Full `scipy.signal`: 15916 passed, 4 skipped, 0 failed.

#### AI Generation Disclosure

Tool used: Claude Code (Claude Opus 5).

AI generated the code in this PR: the validation added to `FIRsepsym2d`, the
docstring change, and the new and updated tests. It also derived the bound and ran
the verification described above, and it grammar-edited the text of this
description.

I picked the issue, chose the exact bound over the more conservative one the TODO
suggested, reviewed every line of the diff, and I take full responsibility for it.

### CODE REVIEW

**Comment by tekinertekin** [comment]
Gentle ping - this has been open for eight days with all 57 CI checks green.
It is one commit, +41/-43 over three files.

@ilayn you noted on gh-24681 that the overflow was suppressed in the ASAN config
"until someone fixes it". This PR removes that suppression, together with the
in-source TODO asking for exactly this ("A proper fix is either to put an input
validation or fix the boundary condition logic for filters larger than the
signal").

One deliberate difference from that TODO: it validates the exact bound rather
than Nh <= N, because len(filt) == n + 1 is still in bounds and agrees with a
mirror-symmetric reference convolution, so the stricter bound would have
rejected working input. Happy to change the approach or the error wording if
you would rather it went another way.

File: scipy/signal/_splinemodule.cc [hunk 1]
```diff
@@ -91,14 +91,15 @@ static PyObject *FIRsepsym2d(PyObject *NPY_UNUSED(dummy), PyObject *args) {
...
-   * logic for filters larger than the signal.
-   *
+  /* _fir_mirror_symmetric reflects an out-of-range index only once, so it needs
+   * at least 2*(Nh/2) samples along the axis it filters. Shorter axes make it
+   * read, and for Nh/2 > N also write, outside the arrays. Lifting the
+   * restriction would require repeated reflection in the boundary sections.
    */
+  if (N < 2 * (PyArray_DIMS(a_hrow)[0] >> 1) || M < 2 * (PyArray_DIMS(a_hcol)[0] >> 1)) {
+    PYERR("hrow must not be longer than image.shape[1] + 1, and hcol must not "
+          "be longer than image.shape[0] + 1");
```
Comment (tylerjereddy) [inline scipy/signal/_splinemodule.cc]: Should we perhaps use `input` instead of `image` here, given that the parameter is called `input` and we're in `scipy.signal` rather than i.e., `ndimage`?

**Review by tylerjereddy — COMMENTED** [review]
It can take a while for the core team to get around to a review--there are a lot of AI-assisted PRs/issues now.

I took a pretty careful look here, despite not being a `signal` regular. Overall, this does look like a step in the right direction, and the regression test changes are all apparently fixing up tests that were unconditionally `xfail`ed anyway, so the risk here seems low even if we don't fully understand all the details under the hood.

I checked locally on ARM Mac on this branch that if I revert your source patch the adjusted `test_sepfir2d_filter_too_long()` does indeed fail. The lack of an ASAN issue after the suppression removal is also a good sign.

So, I think I'd be in favor or merging here, with a preference for a `signal` regular to take a quick look (but maybe not mandatory given risk analysis above).

File: scipy/signal/tests/test_bsplines.py [hunk 2]
```diff
@@ -324,22 +324,32 @@ def test_strided(self, dtyp, xp):
...
+        ((1, 8), 3, 3),    # axis 0 too short even for a 3-tap filter
+        ((2, 2), 7, 7),    # short enough that the output pointer also ran out
+    ])
+    def test_sepfir2d_filter_too_long(self, shape, nrow, ncol, xp):
+        # gh-24681: the boundary sections reflect an out-of-range index only
+        # once, so an axis shorter than 2*(len(filt)//2) used to read, and for
+        # the last case above also write, outside the arrays instead of raising.
+        image = np.ones(shape)
+        with pytest.raises(ValueError, match="must not be longer"):
+            signal.sepfir2d(image, np.ones(nrow), np.ones(ncol))
```
Comment (j-bowhay) [inline scipy/signal/tests/test_bsplines.py]: Here and elsewhere the `xp` fixture should be used in place of `np`, see https://docs.scipy.org/doc/scipy/dev/api-dev/array_api.html#adding-tests

**Review by j-bowhay — DISMISSED** [review]

File: scipy/signal/tests/test_bsplines.py (hunk 2 above)
Comment (tekinertekin) [reply scipy/signal/tests/test_bsplines.py]: Thanks a lot for review @j-bowhay  - converted both tests I added to build their arrays through `xp`
(`xp.ones`, `xp.arange`, `xp.full`, `xp.sum`).

File: scipy/signal/_splinemodule.cc (hunk 1 above)
Comment (tekinertekin) [reply scipy/signal/_splinemodule.cc]: Thanks a lot for the careful look @tylerjereddy  , and for checking the revert locally.

Changed it to `input`.

File: scipy/signal/_splinemodule.cc [hunk 3]
```diff
@@ -91,14 +91,15 @@ static PyObject *FIRsepsym2d(PyObject *NPY_UNUSED(dummy), PyObject *args) {
...
-   * A proper fix is either to put an input validation or fix the boundary condition
-   * logic for filters larger than the signal.
-   *
+  /* _fir_mirror_symmetric reflects an out-of-range index only once, so it needs
+   * at least 2*(Nh/2) samples along the axis it filters. Shorter axes make it
+   * read, and for Nh/2 > N also write, outside the arrays. Lifting the
+   * restriction would require repeated reflection in the boundary sections.
    */
+  if (N < 2 * (PyArray_DIMS(a_hrow)[0] >> 1) || M < 2 * (PyArray_DIMS(a_hcol)[0] >> 1)) {
+    PYERR("hrow must not be longer than input.shape[1] + 1, and hcol must not "
```
Comment (rgommers) [inline scipy/signal/_splinemodule.cc]: It'd be better to report only the failing dimension here.

**Review by rgommers — APPROVED** [review]
Thank you @tekinertekin, this is great - it solves a long-standing problem, and this function has been crashing on Windows in CI recently, so I'm hopeful that it'll solve that as well.

The changes are all correct. I pushed two minor additions, one to improve the error message, and an extra test to test correctness of long filters and not just memory-safeness.

**Comment by rgommers** [comment]
Adding the backport label as well, that will help the next person running their stack under ASan.

**Comment by rgommers** [comment]
This turned all red due to a mistake in another PR, reverted now (xref gh-26085). Re-running CI.

**Comment by rgommers** [comment]
Let's give this a go - thanks again @tekinertekin & reviewers!

**Comment by tekinertekin** [comment]
Thanks a lot everyone for reviewing and approving @rgommers @j-bowhay @tylerjereddy
